### PR TITLE
tensorflow: add a binary wheel release for Darwin

### PR DESCRIPTION
**Pull request by abbradar**
###### Motivation for this change

Re-add Tensorflow on Darwin, albeit in binary form. Closes #34420.

###### Things done

- [ ] Tested using sandboxing ([nix.useSandbox](http://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS, or option `build-use-sandbox` in [`nix.conf`](http://nixos.org/nix/manual/#sec-conf-file) on non-NixOS)
- Built on platform(s)
   - [ ] NixOS
   - [ ] macOS
   - [ ] other Linux distributions
- [ ] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [ ] Tested compilation of all pkgs that depend on this change using `nix-shell -p nox --run "nox-review wip"`
- [ ] Tested execution of all binary files (usually in `./result/bin/`)
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

---

Can someone on Darwin check if this works? Or maybe one knows how can you spin up a Darwin VM on an AMD processor?

### CODE REVIEW

**Comment by abbradar** [comment]
@grahamc can I specify explicitly what platforms do I want to build on with @GrahamcOfBorg ? I can just run `build` and hope Tensorflow on Linux would be taken from binary cache but I don't want to risk running the full build :D

**Comment by lukeadams** [comment]
@abbradar [python36 builds but errors in Keras due to lack of absl (nixpkgs: `pythonPackages.absl-py`)](https://gist.github.com/lukeadams/02675dac3c8d109ccd1ad06f3f805898)

[python27 build fails with this error](https://gist.github.com/lukeadams/4d011e0a944b6ee212b4782745819b6e) Looks like `pythonPackages.enum-compat` in nixpkgs would solve it (Or just adding `python27Packages.enum34` on python2.x only)

I'll have more time to test tomorrow

**Comment by grahamc** [comment]
You can't but there is not much risk. Do it anyway is say! :)

**Comment by abbradar** [comment]
@GrahamcOfBorg build python2.pkgs.tensorflow python3.pkgs.tensorflow

**Comment by abbradar** [comment]
@GrahamcOfBorg build python2.pkgs.tensorflow python3.pkgs.tensorflow

**Comment by abbradar** [comment]
@lukeadams Try this one when you have time.

**Comment by abbradar** [comment]
@grahamc BTW, why did the bot attach "cleanup" label to this?

File: pkgs/development/python-modules/tensorflow/bin.nix [hunk 1]
```diff
@@ -0,0 +1,67 @@
...
+, cudaSupport ? false
+}:
+
+# tensorflow is built from a downloaded wheel because the source
+# build doesn't yet work on Darwin.
+
+buildPythonPackage rec {
+  pname = "tensorflow";
+  version = "1.5.0";
+  name = "${pname}-${version}";
```
Comment (FRidh) [inline pkgs/development/python-modules/tensorflow/bin.nix]: The `name` attribute is added by `buildPython*` and should therefore be removed.

File: pkgs/development/python-modules/tensorflow/bin.nix [hunk 2]
```diff
@@ -0,0 +1,67 @@
...
+                 ++ lib.optionals (pythonOlder "3.4") [ backports_weakref enum34 ]
+                 ++ lib.optional (pythonOlder "3.6") tensorflow-tensorboard;
+
+  # tensorflow depends on tensorflow_tensorboard, which cannot be
+  # built at the moment (some of its dependencies do not build
+  # [htlm5lib9999999 (seven nines) -> tensorboard], and it depends on an old version of
+  # bleach) Hence we disable dependency checking for now.
+  installFlags = lib.optional isPy36 "--no-dependencies";
+
+  doCheck = false;
```
Comment (FRidh) [inline pkgs/development/python-modules/tensorflow/bin.nix]: Why are the tests disabled? Please include a comment.

**Review by FRidh — CHANGES_REQUESTED** [review]

**Comment by lukeadams** [comment]
@abbradar [py27 still has issues when Keras tries to load it](https://gist.github.com/lukeadams/56550ec2b07a333a6fc147156ec403b6).
```
install_name_tool -id "$out/lib/python2.7/site-packages/tensorflow/python/_pywrap_tensorflow_internal.so" "$out/lib/python2.7/site-packages/tensorflow/python/_pywrap_tensorflow_internal.so"
```
should fix the error. Testing right now...


[py26 works though!](https://gist.github.com/lukeadams/3d96c9e435ccfb6eacfec6725320c4f1)

**Comment by abbradar** [comment]
@lukeadams Wait, did you made a mistake or can Darwin use `.so`s too?

**Comment by lukeadams** [comment]
@abbradar Yup, .so on darwin is the same as .dylib

**Comment by lukeadams** [comment]
Hmm this is interesting...
```
  postFixup = lib.optionalString (!isPy3k) ''
    install_name_tool -id "$out/lib/python2.7/site-packages/tensorflow/python/_pywrap_tensorflow_internal.so" \
      "$out/lib/python2.7/site-packages/tensorflow/python/_pywrap_tensorflow_internal.so"

    install_name_tool -change "@rpath/libtensorflow_framework.so" \
      "$out/lib/python2.7/site-packages/tensorflow/libtensorflow_framework.so" \
      "$out/lib/python2.7/site-packages/tensorflow/python/_pywrap_tensorflow_internal.so"
  '';
```
Which changes `otool -L <outpath>/lib/python2.7/site-packages/tensorflow/python/_pywrap_tensorflow_internal.so` output from
```
/nix/store/0vbwkg50jirxlipi69qr631bnwj9qjjw-python2.7-tensorflow-1.5.0/lib/python2.7/site-packages/tensorflow/python/_pywrap_tensorflow_internal.so:
	@rpath/_pywrap_tensorflow_internal.so (compatibility version 0.0.0, current version 0.0.0)
	@rpath/libtensorflow_framework.so (compatibility version 0.0.0, current version 0.0.0)
	/usr/lib/libSystem.B.dylib (compatibility version 1.0.0, current version 1252.0.0)
	/System/Library/Frameworks/IOKit.framework/Versions/A/IOKit (compatibility version 1.0.0, current version 275.0.0)
	/System/Library/Frameworks/CoreFoundation.framework/Versions/A/CoreFoundation (compatibility version 150.0.0, current version 1450.15.0)
	/System/Library/Frameworks/Security.framework/Versions/A/Security (compatibility version 1.0.0, current version 58286.31.2)
```
to 
```
/nix/store/vq1s9dzws7y4648gw4m6lfq1qmzf41px-python2.7-tensorflow-1.5.0/lib/python2.7/site-packages/tensorflow/python/_pywrap_tensorflow_internal.so:
	/nix/store/vq1s9dzws7y4648gw4m6lfq1qmzf41px-python2.7-tensorflow-1.5.0/lib/python2.7/site-packages/tensorflow/python/_pywrap_tensorflow_internal.so (compatibility version 0.0.0, current version 0.0.0)
	/nix/store/vq1s9dzws7y4648gw4m6lfq1qmzf41px-python2.7-tensorflow-1.5.0/lib/python2.7/site-packages/tensorflow/libtensorflow_framework.so (compatibility version 0.0.0, current version 0.0.0)
	/usr/lib/libSystem.B.dylib (compatibility version 1.0.0, current version 1252.0.0)
	/System/Library/Frameworks/IOKit.framework/Versions/A/IOKit (compatibility version 1.0.0, current version 275.0.0)
	/System/Library/Frameworks/CoreFoundation.framework/Versions/A/CoreFoundation (compatibility version 150.0.0, current version 1450.15.0)
	/System/Library/Frameworks/Security.framework/Versions/A/Security (compatibility version 1.0.0, current version 58286.31.2)
```

Additionally, `nm -g <outpath>/lib/python2.7/site-packages/tensorflow/python/_pywrap_tensorflow_internal.so | grep _PyUnicodeUCS2_AsASCIIString` shows that the symbol does exist in the file ...

**Comment by abbradar** [comment]
@lukeadams Ugh, I guess Tensorflow + Python 2 + Darwin is not an option then until we figure out how to make Bazel build there. I'm not sure how did it work before...

**Comment by lukeadams** [comment]
Yup py36 worked without modification

**Comment by abbradar** [comment]
@GrahamcOfBorg build python3.pkgs.tensorflow

**Comment by grahamc** [comment]
@abbradar 

```dif
-    platforms = with platforms; if cudaSupport then linux else linux ++ darwin;
+    platforms = platforms.linux; 
```

This counts as a removed package, which implies cleanup.

**Comment by lukeadams** [comment]
@abbradar I just got bazel building on Darwin! I'll have a pr up soon. The main issue was BUILD files occasionally invoking CC instead of CXX for c++ files on darwin. Buuutt that doesn't mean it's working well enough to build tf, which I'm testing right now.
screen shot 2018-03-16 at 6 32 11 pm

**Comment by abbradar** [comment]
@lukeadams Great! I'll wait a bit with merging this then, let's see if you have some luck.

**Comment by lukeadams** [comment]
@abbradar I'll upload my commits tomorrow morning, but tf isn't buildable with it unfortunately. It successfully compiles everything, but fails to link at the very end (I won't have time to look into it until next weekend). 😢

**Comment by abbradar** [comment]
@lukeadams Wow, much kudos! I'll review the branch after cleanup, it already looks promising.

**Comment by abbradar** [comment]
@GrahamcOfBorg build python3.pkgs.tensorflow python2.pkgs.tensorflow

**Comment by abbradar** [comment]
@GrahamcOfBorg eval
